File: python/백준/자료구조/힙/1715 카드 정렬하기.py
import heapq
n = int(input())

cards = []
result = 0
for _ in range(n):
    heapq.heappush(cards, int(input()))

while len(cards) > 1:
    first = heapq.heappop(cards)
    second = heapq.heappop(cards)
    
    result += (first + second)
    heapq.heappush(cards, first + second)

print(result)

File: cpp/백준/그리디/1715 카드 정렬하기.cpp
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    priority_queue<long long, vector<long long>, greater<long long>> cards;
    for (int i = 0; i < n; ++i) {
        long long x;
        cin >> x;
        cards.push(x);
    }

    long long result = 0;
    while (cards.size() > 1) {
        long long first = cards.top(); cards.pop();
        long long second = cards.top(); cards.pop();

        result += (first + second);
        cards.push(first + second);
    }

    cout << result << "\n";
    return 0;
}
